benchmark/Streamly/Benchmark/Data/Array.hs

```diff
@@ -4,99 +4,191 @@
-        , benchIOSrc "write . intFromTo" Ops.sourceIntFromToFromStream
-       , benchIOSrc "fromList . intFromTo" Ops.sourceIntFromToFromList
-        , benchIOSrc "writeN . unfoldr" Ops.sourceUnfoldr
-        , benchIOSrc "writeN . fromList" Ops.sourceFromList
-        -- , benchPureSrc "writeN . IsList.fromList" Ops.sourceIsList
-        -- , benchPureSrc "writeN . IsString.fromString" Ops.sourceIsString
-        , testStr `deepseq` (bench "read" $ nf Ops.readInstance testStr)
-        , benchPureSink "show" Ops.showInstance
+o_1_space_generation :: Int -> [Benchmark]
+o_1_space_generation value =
+    [ bgroup
+        "generation"
+        [ benchIOSrc "writeN . intFromTo" (Ops.sourceIntFromTo value)
+#ifndef DATA_SMALLARRAY
+        , benchIOSrc "write . intFromTo" (Ops.sourceIntFromToFromStream value)
+#endif
+        , benchIOSrc
+              "fromList . intFromTo"
+              (Ops.sourceIntFromToFromList value)
+        , benchIOSrc "writeN . unfoldr" (Ops.sourceUnfoldr value)
+        , benchIOSrc "writeN . fromList" (Ops.sourceFromList value)
+#ifdef MEMORY_ARRAY
+        , benchPureSrc "writeN . IsList.fromList" (Ops.sourceIsList value)
+        , benchPureSrc
+              "writeN . IsString.fromString"
+              (Ops.sourceIsString value)
+#endif
+#ifndef DATA_PRIM_ARRAY
+#ifdef DATA_SMALLARRAY
+        , let testStr =
+                  "fromListN " ++
+                  show (value + 1) ++
+                  "[1" ++ concat (replicate value ",1") ++ "]"
+#else
+        , let testStr = mkListString value
+#endif
+           in testStr `deepseq` (bench "read" $ nf Ops.readInstance testStr)
+#endif
+        , benchPureSink value "show" Ops.showInstance
         ]
-      , bgroup "elimination"
-        [ benchPureSink "id" id
-        , benchPureSink "==" Ops.eqInstance
-        , benchPureSink "/=" Ops.eqInstanceNotEq
-        , benchPureSink "<" Ops.ordInstance
-        , benchPureSink "min" Ops.ordInstanceMin
+    ]
+
+o_1_space_elimination :: Int -> [Benchmark]
+o_1_space_elimination value =
+    [ bgroup "elimination"
+        [ benchPureSink value "id" id
+        , benchPureSink value "==" Ops.eqInstance
+        , benchPureSink value "/=" Ops.eqInstanceNotEq
+        , benchPureSink value "<" Ops.ordInstance
+        , benchPureSink value "min" Ops.ordInstanceMin
+#ifdef MEMORY_ARRAY
         -- length is used to check for foldr/build fusion
-        -- , benchPureSink "length . IsList.toList" (length . GHC.toList)
-        , benchIOSink "foldl'" Ops.pureFoldl'
-        , benchIOSink "read" (S.drain . S.unfold A.read)
-        , benchIOSink "toStreamRev" (S.drain . A.toStreamRev)
+        , benchPureSink value "length . IsList.toList" (length . GHC.toList)
+#endif
+        , benchIOSink value "foldl'" Ops.pureFoldl'
+        , benchIOSink value "read" Ops.unfoldReadDrain
+        , benchIOSink value "toStreamRev" Ops.toStreamRevDrain
+#ifndef DATA_PRIM_ARRAY
```

Comment:
You could use `#if !defined(DATA_PRIM_ARRAY) && defined(DEVBUILD)`.

Reply:
I didn't rewrite on purpose. I'm thinking of the constants that define what type of array this module is aimed for as top-level conditionals